### ISSUE TITLE
Support patterns following & 

### ISSUE THREAD

**Issue by SevereOverfl0w** [noprompt/meander#91]
Eg `[& ?xs ?x]` when matched in `[1 2 3]` would bind ?xs to `[1 2]` and ?x to `3`.

**Comment by timothypratley** [comment]
This would be very useful for maps where we are forced to use key-value pairs :)

For vectors though, I don't see the advantage as we can already more succinctly express this use case:

```clojure
 (m/match [1 2 3]
          [!xs ... ?x]
          {:xs !xs
           :x ?x})
#_#_=> {:xs [1 2], :x 3}
```

In Clojure only 1 symbol may follow `&`.

```clojure
(defn f [x & y z]
  1)
```
Is an error:
(z) - failed: Extra input at: [:fn-tail :arity-1 :params] spec: :clojure.core.specs.alpha/param-list

I think Meander should follow this convention and make multiple trailers an error.

A consistency with Clojure syntax and still following patterns after & can be achieved by treating the thing after & as either a variable or a sequence. This is consistent with how Clojure always binds a sequence.

```
(m/match [1 2 3]
          [& (!xs ...)]
          !xs)
```
^^ I think support this style would be consistent but doesn't help with vectors

It does, however, solve some serious issues with map matching; preserving order and symmetry.

**Comment by noprompt** [comment]
There are some instances where this would be useful for substitution, however, I suppose I should probably demand some more examples and argument before I sign on off it on it. 😄

**Comment by timothypratley** [comment]
TLDR I can see the value in this now

`[1 2 & ?rest]` is very similar to `[1 2 . !rest ...]` so let's just focus on the differences:
* The former is indeed shorter and looks more like "Clojure"
* `?rest` has a different meander name "variable" to `!rest` which is a "memory variable"
* It's interesting to think about what type they should be, more on that later!
* `...` has convenience for some situations; consider `[1 2 . !x !y ...]` is effectively partitioning '& ?rest`
* Some cases where symmetry is a consideration:
```clojure
(s/rewrite [1 2 & ?rest] [1 2 & ?rest :terminator])
(s/rewrite [1 2 & ?rest :terminator] [1 2 & ?rest])
```
Much more visually appealing than forcing people to use `...`

```clojure
[& _ & ?xs :separator & ?ys :another-separator & _]
[_ ... !xs ... :separator . !ys ... :another-separator _ ...]
[& _ !xs ... :separator . !ys ... :another-separator & _]
(scan !xs :separator !ys :another-separator)
```
All of these seem perfectly reasonable but qualitatively I have to say I like the 3rd style where `& _` is used to bookend the expression. Mixing `&` and `...` is fine.

I raised the objection "Clojure doesn't allow this" but that does not hold up under scrutiny... meander is not subject to Clojure syntax rules and must explicitly break from them to provide any functionality, so please disregard that as lazy thinking on my part. The departure is additive and no conflicts arise.

Regarding types, I opened a new issue #101 as it is orthogonal to this particular consideration.

**Comment by timothypratley** [comment]
Another consideration here is that `&` would effectively provide grouping semantics (a good thing IMO) because you could write things like `[& [1 2 3 :as ?group1] & [4 5 6 :as ?group2]]`
which got me thinking... another way to express this notion is `[1 2 3 ~@(4 5 6 :as ?z) 7 8 9]`

I'm not sure what the implication is here but to my eyes `&` and `~@` have the same meaning.

**Comment by noprompt** [comment]
> I'm not sure what the implication is here but to my eyes `&` and `~@` have the same meaning.

Pretty much but the main difference is that `& <pattern>` in this case applies substitution.

```clj
[1 2 3 & [!1 !2 ... & ?rest]]
```

**Comment by noprompt** [comment]
This will be available in the next release of the library after #180 is merged. 😄 

Sorry it took me so long to get around to that.
